synapse/api/auth.py

```diff
@@ -119,6 +267,202 @@ async def check_user_in_room(
+
+        # We currently require the user is a "moderator" in the room. We do this
+        # by checking if they would (theoretically) be able to change the
+        # m.room.canonical_alias events
+
+        power_level_event = (
+            await self._storage_controllers.state.get_current_state_event(
+                room_id, EventTypes.PowerLevels, ""
+            )
+        )
+
+        auth_events = {}
+        if power_level_event:
+            auth_events[(EventTypes.PowerLevels, "")] = power_level_event
+
+        send_level = event_auth.get_send_level(
+            EventTypes.CanonicalAlias, "", power_level_event
+        )
+        user_level = event_auth.get_user_power_level(
+            requester.user.to_string(), auth_events
+        )
+
+        return user_level >= send_level
+
+
+class AccessTokenMixin:
+    @staticmethod
+    def has_access_token(request: Request) -> bool:
+        """Checks if the request has an access_token.
+
+        Returns:
+            False if no access_token was given, True otherwise.
+        """
+        # This will always be set by the time Twisted calls us.
+        assert request.args is not None
+
+        query_params = request.args.get(b"access_token")
+        auth_headers = request.requestHeaders.getRawHeaders(b"Authorization")
+        return bool(query_params) or bool(auth_headers)
+
+    @staticmethod
+    def get_access_token_from_request(request: Request) -> str:
+        """Extracts the access_token from the request.
+
+        Args:
+            request: The http request.
+        Returns:
+            The access_token
+        Raises:
+            MissingClientTokenError: If there isn't a single access_token in the
+                request
+        """
+        # This will always be set by the time Twisted calls us.
+        assert request.args is not None
+
+        auth_headers = request.requestHeaders.getRawHeaders(b"Authorization")
+        query_params = request.args.get(b"access_token")
+        if auth_headers:
+            # Try the get the access_token from a "Authorization: Bearer"
+            # header
+            if query_params is not None:
+                raise MissingClientTokenError(
+                    "Mixing Authorization headers and access_token query parameters."
+                )
+            if len(auth_headers) > 1:
+                raise MissingClientTokenError("Too many Authorization headers.")
+            parts = auth_headers[0].split(b" ")
+            if parts[0] == b"Bearer" and len(parts) == 2:
+                return parts[1].decode("ascii")
+            else:
+                raise MissingClientTokenError("Invalid Authorization header.")
+        else:
+            # Try to get the access_token from the query params.
+            if not query_params:
+                raise MissingClientTokenError()
+
+            return query_params[0].decode("ascii")
+
+
+class LegacyAuth(BasePermissions, AccessTokenMixin):
```

Comment:
I see that a later commit *does* this. As this is already a very large PR, it could help to do this up-front in the initial commits (I don't know how terrible that rebase would be unfortunately).